### ISSUE TITLE
Trouble with output device sound capture

### ISSUE THREAD

**Issue by kailax** [sskodje/ScreenRecorderLib#75]
I'm using this library to make my own screen capture software, i like the features and the results are pretty decent! however i'm having issues with the sound recording options. When i enable output device recording, i can't hear the computer sound in the recorded file. Not only that, when i enable both output and input audio, the input audio from my laptop microphone gets all choppy or most of the time is not audible at all. I tried tweaking on the AudioOptions parameters, but its no use, this happens even if i set everything to the default values and use the default audio capture devices. When i record with just the input (microphone) it works fine

**Comment by Willy-Kimura** [comment]
True, I've likewise received the same issue when I/O audio recording options are both enabled.

**Comment by hi2u** [comment]
I've just started using this library today, and also have an audio issue, but it's the opposite for me... recording my speakers (output) works, but not my microphone (input).  Not sure if related, but maybe a recent change affected something in common.

Here's my current code to init the recording:

```csharp
_rec = Recorder.CreateRecorder(new RecorderOptions {
	AudioOptions = new AudioOptions
	{
		IsAudioEnabled = true,
                // --------------------------------------------------------------------------
		// Speakers: "Speakers (Realtek High Definition Audio)"
		// This works:
		IsOutputDeviceEnabled = true,
		AudioOutputDevice = "{0.0.0.00000000}.{fe5ea803-5ced-476e-8eba-f6bea3b2311d}",
		// --------------------------------------------------------------------------
		// --------------------------------------------------------------------------
		// Microphone: "Headset Microphone (2- Microsoft LifeChat LX-3000 )"
		// This does NOT work - my program just silently closes/crashes without throwing an exception or anything
		IsInputDeviceEnabled = true,
		AudioInputDevice = "{0.0.1.00000000}.{05181496-38b1-491d-b789-89bedc2a20b5}",
		
	}
});
```

I've been comparing results by comment out specific audio devices.  Here's the 3 scenarios I'm testing, and the result of each:

# Working

* Scenario #1: "Speakers" code above enabled + "Microphone" code commented out:
    *  ✔️ Result: All good.

# Fails

* Scenario #2: "Speakers" code above commented out + "Microphone" code enabled
* -and also-
* Scenario #3: both enabled
    * ❌Result: Fails
     * When recording starts, my winforms program freezes for a few seconds, and then the window/program just silent closes/crashes
     * No exception is thrown, nor is any kind of error message shown anywhere
     * The output .mp4 file is created, but just stays empty (0 bytes)

**Comment by sskodje** [comment]
I have reproduced the issues, and i am working on a fix.

**Comment by kailax** [comment]
I'm testing version 2.4.0, and i'm sharing results here:

- Audio Output set to **Disabled** results in Recording files to show up corrupted (file size of only 79 bytes, codec information empty)

- Audio Input  set to **Enabled** results in software to crash almost immediately after i begin the recording, this is the output:

> El subproceso 0x22d8 terminó con código 0 (0x0).
El programa '[2912] MYPROGRAM.exe' terminó con código -1073741819 (0xc0000005) 'Access violation'.

- With Audio Recording **Disabled** there are no issues, also no issues with Audio Recoding **Enabled** but Output and Input **Disabled**

**Comment by sskodje** [comment]
The issues with sound recording should be fixed with version 2.5.0. If there still are issues, let me know and i'll reopen the issue.

**Comment by kailax** [comment]
Hi, thank you very much. I tested just now and this are the results i got:

- Audio Input **Enabled** with Audio Output **Disabled**: No issues, recorded video has audio from the microphone and video is fine.

- Audio Input **Enabled** with Audio Output **Enabled**: Same issue as before, for some reason the recording audio is getting chopped and is not even audible. I analyzed the audio with Adobe Audition and after amplifying audio by +30db i could hear the audio but it's like a constant stutter (i uploaded sample audio to my drive here: https://drive.google.com/file/d/117RQ0vyitUI2oj4C-w5sAvXVs3MeHr_w/view?usp=sharing).

- Audio Input **Disabled** with Audio Output **Enabled**: Same issue as above, i think the issue has something to do with the way the loopback is being recorded or resampled

And of course if i Disable the audio there are no issues and recording is fine

**Comment by Willy-Kimura** [comment]
Thanks for the update.

I tested the latest update `2.5.0` with I/O audio enabled and it's working well now.
Probably @kailax you can share your system specifications.

**Comment by kailax** [comment]
Thanks @Willy-Kimura, i was also thinking i may have a hardware compatibility issue. I'm testing in a Lenovo Ideapad 330s with AMD Ryzen 5 chipset, the audio drivers are Realtek drivers ( for input and output). This hardware supports Stereo Mix recording, i have recorded computer output audio with software like Adobe Audition using that feature, but i thought this wouldn't be necessary in all cases since software like OBS supports recording audio output out of the box

https://www.loom.com/share/83c891fd2fa34d228ba7853b69f7f4a1

**Comment by sskodje** [comment]
Do you have your sound output configured for more than two channels, like 5.1 or 7.1? What's your audio output bit depth and sample rate?

**Comment by kailax** [comment]
This is my audio out configuration:

- **Bit depth**: 16 bits
- **Sample rate**: 48000 Hz

(This config i cannot change)

I also disabled the Windows 10 audio enhancements and Hardware acceleration, but it was the same result. If i activate the Stereo Mix and record using that as the input the audio is recorded correctly without issues.

EDIT: I did some tests with CSCore, recording audio loopback (WASAPI) and the audio recorded was audible but slower (like 0.3x normal speed).

**Comment by YongbeomBahg** [comment]
@kailax 
I am not familiar with audio but as I know, this is a kind of stereo mix
MIC in sound should not out from speaker to prevent echo sound
What about use USB interface head set then you may acquire your purpose

**Comment by sskodje** [comment]
I'm not too knowledgeable about audio myself unfortunately, but i don't think it should be necessary with any special driver features. It should work with default windows sound drivers for realtek or any other standard audio chips.

If recorded audio is slower or faster than it should be, that sounds like a problem with the sample rate. As an example, if you record something that is 22khz as 44.1khz without resampling, it will play back at 1/2 the normal speed. I don't see how this should be a problem though, since your settings are 16 bit 48khz, which is natively supported and does not require or trigger resampling.

**Comment by kailax** [comment]
> @kailax
> I am not familiar with audio but as I know, this is a kind of stereo mix
> MIC in sound should not out from speaker to prevent echo sound
> What about use USB interface head set then you may acquire your purpose

@YongbeomBahg Well that may solve my problem, but that is not the issue (i can already solve the issue on this device by enabling the stereo mix). I'm coding a recording software and it should be able to record in most environments without a lot of tweaking. 

@sskodje Yes, exactly i also think the issue is related to the audio loopback sampling/resampling process. I will keep experimenting, i was thinking on recording the audio loopback separately with CSCore and the mixing it with the video recording, but the resulting audio also has issues. I will record and upload audio using the two libraries if you think comparing them can help identifying the issue

**Comment by jpcanepa** [comment]
I'm running into this as well, with, at least 2.6.0. It is weird that it heppens only when audio from _both_ inputs is played, but if either is played separately (even though both interfaces are configured for capture) it seems ok. I'm investigating this issue too, but if anyone has pointers, please update this issue.

**Comment by jpcanepa** [comment]
After running a number of tests, it seems that the problem (at least for me) is exclusively if I set the AudioChannels property to mono. @kailax are you recording stereo or mono?

**Comment by jpcanepa** [comment]
Ok, for whatever it is worth, I think I found out the source of my problems. 

The `MixAudio` function takes two buffers of PCM sample data, and tries to sum them up to mix them. The problem is that if one of the buffers is longer, then it outputs all the data of the longer buffer, mixing in the shorter buffer. That means that, for example, if one of the inputs is silent, the mixed audio will have a discontinuity.

The solution I found was to return only the mixed portion, and store the leftover of the first/second buffer for the next call, prepending it to the incoming data. That way the audio dropouts go away.

I'll try to push a PR.

**Comment by sskodje** [comment]
Hmm. So you mean that the result from [this line](https://github.com/sskodje/ScreenRecorderLib/blob/64773edf2994eb241a532f471bade780c83ba408/ScreenRecorderLib/internal_recorder.cpp#L887) has fewer bytes than it asks for, so the two vectors have a difference in length?

**Comment by jpcanepa** [comment]
@sskodje [this](https://github.com/descriptinc/ScreenRecorderLib/pull/5/files) is what I did. It is kind of messy, so I don't feel like turning it into a proper PR, but perhaps you can adapt it to a more idiomatic implementation.

**Comment by jpcanepa** [comment]
DANG. While writing the previous comment I think I found a bug in my own implementation. In [this line](https://github.com/descriptinc/ScreenRecorderLib/pull/5/files#diff-c65ffae22093bdbe6a762fde58ec5a31832d00d0dbeb7d71313733336b585a86R181) the default value for the `isLastCall` param should be `false` not `true`. The rest of the argument still stands, though.

**Comment by jpcanepa** [comment]
Phew, mi previous comment is correct, but the bug has no effect because every call to `MixAudio` comes from to `GrabAudioFrame` that passes in the correct value.

**Comment by sskodje** [comment]
thanks, i'll have a look

**Comment by sskodje** [comment]
I made some fixes thanks your your feedback in 506c4c39992a8b99e3662adf7e964d6cee4a179c, so i believe this issue should be fixed. Thanks again!
